### ISSUE TITLE
BAD error in R 3.4.0

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#1222]
I upgraded to R 3.4.0 yesterday and now I see a very bad error (below).

```R
> library(oce)
Loading required package: gsw
> data(ctd)
> plot(ctd)

 *** caught segfault ***
address 0x18, cause 'memory not mapped'

Traceback:
 1: .Call("geoddist", as.double(latitude1), as.double(longitude1),     as.double(latitude2), as.double(longitude2), as.double(a),     as.double(f))
 2: geodDist(mcoastline[["longitude"]], mcoastline[["latitude"]],     mean(x[["longitude"]], na.rm = TRUE), mean(x[["latitude"]],         na.rm = TRUE))
 3: .local(x, ...)
 4: plot(ctd)
 5: plot(ctd)

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
```

**Comment by dankelley** [comment]
I added some print statements to `src/geod.c` and the problem went away. 😒

> Where art thou, Muse, that thou forget'st so long
> To speak of that which gives thee all thy might?
> -- Shakespeare Sonnet 100

**Comment by dankelley** [comment]
Oh, this is weird. I removed my debugging lines, one by one, and the problem did not come back. Now I am at the exact same source as yesterday (verified by git) and ... it works fine, with no segfault, on my test:
```R
R -e "library(oce);data(ctd);plot(ctd)"
```
so I am left with a bit of a mystery. But this is par for the course with C; the problem may depend on memory pressure from other applications, etc. I will leave this open for a week so that other users will know to add notes here, if they upgrade. (Of course, I cannot even know that it relates to the upgrade; all I can say is that this `geoddist` seems to have been working for years on end, so I suspect this is something new.)

**Comment by richardsc** [comment]
Weird. I'll try some tests on my system (Windows and Linux) to see if I see any issues. But, that involves updating R which also means re-building all the packages I use (at least on Linux ...)

**Comment by dankelley** [comment]
Thanks.  If you know how to turn on memory-fault detection (like `valgrind`), that would help a lot, too.  It can catch some errors of going beyond the length of an array.

C problems are hard to detect, because they can be inconsistent (depending on memory pressure, etc).  We have some code in `src/geod.c` that is fairly new, and I'm going to look it all over for things that relate to memory or the possibility of going past allocated memory.

**Comment by dankelley** [comment]
Maybe you can try first with the existing R, but with `valgrind` turned on.  It's possible this is a bug that's active in the old R, too.

**Comment by richardsc** [comment]
Hm. Too late -- I'm already 95% of the way through the upgrade in both Windows and Linux.

How do you turn `valgrind` on?

**Comment by richardsc** [comment]
Test on Windows 7, R3.4.0, built from commit 56c24900:
```
library(oce)
data(ctd)
plot(ctd)
```
Successful!

**Comment by dankelley** [comment]
I'm now in a virtualbox, trying to remember about turning valgrind on!  I just googled http://www.hep.by/gnu/r-patched/r-exts/R-exts_87.html which makes me recall that in linux (and, I think windows) the compile step need not be altered.  Just use `R -d valgrind --vanilla < test.R`

**Comment by richardsc** [comment]
Same test, same commit, in Linux Mint, also success.

I didn't try with valgrid yet

**Comment by richardsc** [comment]
I can't get valgrind to work using the command above. I get:
```
$ R -d valgrind --vanilla < test.R
*** Further command line arguments ('--vanilla ') disregarded
*** (maybe use 'run --vanilla ' from *inside* valgrind)

/usr/bin/R: line 286: exec: valgrind: not found
```

**Comment by dankelley** [comment]
I think you will need to do an aptget (or whatever) to install valgrind. Maybe http://www.hep.by/gnu/r-patched/r-exts/R-exts_87.html has hints? Or maybe just go to http://www.valgrind.org and install from there?

Meantime I'm swearing at the terminal in linux/virtualbox, which makes me type a quote mark several times to appear, presumably because it's all multilingual and sh1t.

**Comment by dankelley** [comment]
When you get valgrind on there it should look like below. (This is linux on my virtualbox, with an old R)

<img width="871" alt="screen shot 2017-04-23 at 11 18 24 am" src="https://cloud.githubusercontent.com/assets/99469/25314241/aacdb0a0-2816-11e7-898a-d188441d3d3f.png">

**Comment by richardsc** [comment]
Got it.

```
$ R -d valgrind --vanilla < test.R
==11767== Memcheck, a memory error detector
==11767== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==11767== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==11767== Command: /usr/lib/R/bin/exec/R --vanilla
==11767== 

R version 3.4.0 (2017-04-21) -- "You Stupid Darkness"
Copyright (C) 2017 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(oce)
Loading required package: gsw
> data(ctd)
> plot(ctd)
==11767== Conditional jump or move depends on uninitialised value(s)
==11767==    at 0x4F1718B: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F177A5: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F1A4F8: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F1A5F9: GEPolygon (in /usr/lib/R/lib/libR.so)
==11767==    by 0xAE50EBA: ??? (in /usr/lib/R/library/graphics/libs/graphics.so)
==11767==    by 0xAE6608F: ??? (in /usr/lib/R/library/graphics/libs/graphics.so)
==11767==    by 0x4F0ADF0: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F0D071: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F41436: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F49487: Rf_eval (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F4B8BE: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F49655: Rf_eval (in /usr/lib/R/lib/libR.so)
==11767== 
==11767== Conditional jump or move depends on uninitialised value(s)
==11767==    at 0x4F177A8: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F1A4F8: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F1A5F9: GEPolygon (in /usr/lib/R/lib/libR.so)
==11767==    by 0xAE50EBA: ??? (in /usr/lib/R/library/graphics/libs/graphics.so)
==11767==    by 0xAE6608F: ??? (in /usr/lib/R/library/graphics/libs/graphics.so)
==11767==    by 0x4F0ADF0: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F0D071: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F41436: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F49487: Rf_eval (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F4B8BE: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F49655: Rf_eval (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F4C537: ??? (in /usr/lib/R/lib/libR.so)
==11767== 
==11767== Conditional jump or move depends on uninitialised value(s)
==11767==    at 0x4F17123: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F177A5: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F1A4F8: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F1A5F9: GEPolygon (in /usr/lib/R/lib/libR.so)
==11767==    by 0xAE50EBA: ??? (in /usr/lib/R/library/graphics/libs/graphics.so)
==11767==    by 0xAE6608F: ??? (in /usr/lib/R/library/graphics/libs/graphics.so)
==11767==    by 0x4F0ADF0: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F0D071: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F41436: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F49487: Rf_eval (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F4B8BE: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F49655: Rf_eval (in /usr/lib/R/lib/libR.so)
==11767== 
==11767== Conditional jump or move depends on uninitialised value(s)
==11767==    at 0x4F17100: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F177A5: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F1A4F8: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F1A5F9: GEPolygon (in /usr/lib/R/lib/libR.so)
==11767==    by 0xAE50EBA: ??? (in /usr/lib/R/library/graphics/libs/graphics.so)
==11767==    by 0xAE6608F: ??? (in /usr/lib/R/library/graphics/libs/graphics.so)
==11767==    by 0x4F0ADF0: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F0D071: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F41436: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F49487: Rf_eval (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F4B8BE: ??? (in /usr/lib/R/lib/libR.so)
==11767==    by 0x4F49655: Rf_eval (in /usr/lib/R/lib/libR.so)
==11767== 
> 
==11767== 
==11767== HEAP SUMMARY:
==11767==     in use at exit: 170,720,863 bytes in 57,706 blocks
==11767==   total heap usage: 160,504 allocs, 102,798 frees, 807,718,078 bytes allocated
==11767== 
==11767== LEAK SUMMARY:
==11767==    definitely lost: 360 bytes in 1 blocks
==11767==    indirectly lost: 0 bytes in 0 blocks
==11767==      possibly lost: 0 bytes in 0 blocks
==11767==    still reachable: 170,720,503 bytes in 57,705 blocks
==11767==         suppressed: 0 bytes in 0 blocks
==11767== Rerun with --leak-check=full to see details of leaked memory
==11767== 
==11767== For counts of detected and suppressed errors, rerun with: -v
==11767== Use --track-origins=yes to see where uninitialised values come from
==11767== ERROR SUMMARY: 19762 errors from 4 contexts (suppressed: 0 from 0)
```

**Comment by dankelley** [comment]
I'm going to edit this comment to store my tests as I do them.
1. linux virtualbox; R 3.2.3; using valgrind is OK.

**Comment by richardsc** [comment]
Upgrading R in linux is easy if you add the CRAN apt source for your distribution, e.g. add
```
deb http://cran.rstudio.com/bin/linux/ubuntu xenial/
```
to `/etc/apt/sources.list`, and then do an
```
sudo apt-get update
sudo apt-get upgrade
```
to install the latest of everything.

**Comment by dankelley** [comment]
Thanks. I didn't know that `xenial` was the new one.  It took me a while to get to this test because my virtualbox disk got full and I tried increasing it, then gave up and reinstalled linux from scratch etc. So, now I get some of those "definitely lost" problems, but I don't think they mean much. The real problem is when you get things like we had in https://github.com/dankelley/oce/issues/1126

So, I don't know where we stand on this. I still think I'll leave the issue open a week, just in case. But I just don't know if we have a problem or not.

**Comment by richardsc** [comment]
I tried a slightly different test (one that doesn't make the map plot), and got:
```
==14879== 
==14879== HEAP SUMMARY:
==14879==     in use at exit: 119,307,906 bytes in 42,911 blocks
==14879==   total heap usage: 82,924 allocs, 40,013 frees, 394,934,732 bytes allocated
==14879== 
==14879== LEAK SUMMARY:
==14879==    definitely lost: 360 bytes in 1 blocks
==14879==    indirectly lost: 0 bytes in 0 blocks
==14879==      possibly lost: 0 bytes in 0 blocks
==14879==    still reachable: 119,307,546 bytes in 42,910 blocks
==14879==         suppressed: 0 bytes in 0 blocks
==14879== Rerun with --leak-check=full to see details of leaked memory
==14879== 
==14879== For counts of detected and suppressed errors, rerun with: -v
==14879== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
So it still has lots of "definitely lost" leaks, but note that there are no errors.

**Comment by dankelley** [comment]
This is weird. On my work machine, I updated Rstudio and R, and then rebuilt oce (after installing `knitr`, `devtools`, `roxygen2` and `gsw`, plus others that I've forgotten). And then I tried doing the `plot(ctd)` test in Rstudio, and it crashed. Then I tried in R, which is nicer because you get to see more details on the problem, and I see as in the following snapshot.  I'm happy with this because it indicates a problem that's not just on my home machine.  (Note: the docs say this new R needs a certain version of GNU fortran, which of course I don't have on either machine...)

![screen shot 2017-04-24 at 8 39 14 am](https://cloud.githubusercontent.com/assets/99469/25335633/3f709b7e-28ca-11e7-8900-e13289258d47.png)

**Comment by dankelley** [comment]
Hm. I did "clean and rebuild" in RStudio (which forces recompilation of the .c and .f files) and now it works fine.  One thing that's different in R 3.4 is that they handle registration of externally-compiled functions differently. I don't know the details of this, though.  (It has to do with the "register routines" problem that I fixed by adding `src/registerDynamicSymbol.c` in f84f1ab9 on 2017-03-27.)
